### PR TITLE
Openintegrationhub 32

### PR DESCRIPTION
**Pull request by anton-kotenko**
PR that solves issued #32  #24 #26 #28

### CODE REVIEW

File: scripts/build.js
Comment (anton-kotenko) [inline scripts/build.js]: Fixme revert back when integration-content-repository build will be fixed. Details: #93

Comment (falkvoigt) [reply scripts/build.js]: That will remove all Services from the build

Comment (falkvoigt) [reply scripts/build.js]: If you want to remove icr from build make a condition in the check for changes script!

Or remove it from the Object

Comment (anton-kotenko) [reply scripts/build.js]: Sorry, but I don't understand this:
> That will remove all Services from the build

This actually removes only icr service (not all services) from build, as it is broken and can not be built until #93 will be fixed. (example of build logs in travis https://travis-ci.org/openintegrationhub/openintegrationhub/jobs/432142791)

But if you think that it would be better to address this issue in scripts/getChanges.js -- I don't mind and make make it in this way.

Generally I'd prefer to wait until #93 will be fixed before merge this PR, so it won't be required to merge some strange changes and then in few days revert them back, but this depends on plans to fix #93. @falkvoigt  do you know smth about it?

File: scripts/build.js
Comment (anton-kotenko) [inline scripts/build.js]: If you are speaking about this changes, it also should be reverted. But I'd like to discuss build process in general: I don't think that it's good idea to set `latest` tag to any successfully built PR's docker image, because it increases risks that somebody deploys unstable version somewhere.

File: scripts/build.js
Comment (falkvoigt) [reply scripts/build.js]: #93 is fixed already, but the dockerfile needs to be reworkt in icr.

Latest is only set on master builds. Because build and deploy only hits master branches.

I will review the changes tomorrow. Perhabs i missed something on my mobil ;) will come back on it

File: scripts/build.js
Comment (falkvoigt) [reply scripts/build.js]: That latest is only on master branch. 
The deployment will not use latest. The latest tag is just a marker for the latest image not of stable or unstable.
But as i‘m not using it feel free to remove the latest tag ;)

And i tested the Pipeline on iam without Problems.

As i said in the other post i need a big screen to Review that changes. Could be i missed something.

Comment (falkvoigt) [reply scripts/build.js]: Hi,
so i had a look on it. I did not see the hole change you did yesterday. 
Thats fine ! But you disable the Push to DockerHub completely. So we don't get any image to deploy to Cluster. (I just get why you did that) We need to think about a solution here. same for Point 1.

That Latest tag i could understand we could move that tagging to the sage comming up after Deploy in the future (integration Testing not yet implemented).

But: 
1. Why are you disabling my gcloud installation? (just see why you disable it you are not useing gcloud right? but we can't merge that in OIH because we are relating on that installation for deployment)
2. Could you please write the K8S stuff in yaml? that is more common in the Comunity.
3. you added src/ as workspace in yarn could you please name it so every partner could use /src as a lib dir ? (src/backendCommonLib or something)

File: scripts/build.js
Comment (drobiazko) [reply scripts/build.js]: @falkvoigt #93 is open yet and it's blocking this pr. What can we do to proceed?

Comment (falkvoigt) [reply scripts/build.js]: we need to merge the changes @hschmidthh did in his Branch an it's fine. @drobiazko please have a look at my comment below. there are some changes you did in the repo which needs some attention.

File: scripts/build.js
Comment (anton-kotenko) [reply scripts/build.js]: Hi, @falkvoigt.
Let me explain the reason why i did all the changes related to build process: The reason is one and simple: build process starts on every branch (including my pull request, not only master). But build process has next problems, that makes impossible for my PR to pass tests on travis
1. #93 private base image for  integration-content-repository
2. result of every build is pushed into dockerhub with latest tag (even in case of pull request). It's a) dangerous as it increases changes to deploy some incorrect b) there is no access for me to publish 
3. There is deploy phase in all travis scenario. It runs script, that just install gcloud and has commented kubectl commands. But it does not work for me, as it requires some gcloud keys in environment.

So at the moment, there is a problem, that build works only for limited set of users and this actually contradicts open-source work strategy.
Due to reasons written above I had to disable some functionality inside build scripts.

Generally I see two ways to solve all this:
1. I wait until all this issues will be fixed and then merge my PR. Then all this changes won't be required and i return back
2. Merge this PR with all this changes and then fix this issues. 
Your opinion?  

About yaml/json: I prefer json's for next reasons: we are working with javascript, so it's much easier to work with json then with yaml. If you insist on using yaml -- I can transform json descriptors into yaml, but I prefer to do this as separate task.

About src folder: Ok, I'll rename  src/lib to src/backendCommonLib

Comment (falkvoigt) [reply scripts/build.js]: Hi , @anton-kotenko 
the problem is relate to you fork the hole mono repo and also duplicate the travis pipeline. thats why you miss some informations. 

did you read my comment completely? 

1. yes Build breaks related to the Dockerfile fix is done in a different branch needs to be merged

2. i am fine with removing the latest tag for now until we have integration test an Production deployment. But push the build tag is needed for OIH Project.

3. That is related to your fork but OIH relies on this for deployment in test Cluster you can't get rid of it. what you can do , is make a condition relating github Projekt or simple do branching in OIH github instead of forking the hole Project.

We should go with the Community related to Kubernetes in our Github repo. So yaml is prefered Format here. Also a part of the DevOps Guideline which waiting in PR at the moment. 

Over all, the Problem is you forked the Project. We could not merge your PR aslong that commits related to google Cloud and docker push are in there. Which will disable 50% of the Build/deployment Pipeline.

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js:23]: I wanted to suggest to do it with `express Router`, but probably it would be too much for now.

File: scripts/build.js
Comment (anton-kotenko) [reply scripts/build.js]: Sorry, but at this moment:
1) I can not  push new branch into original repo because of lack of privileges, so i have to fork repo, with all consequences (like broken travis pipeline)
2) My build fails in travis without this changes.
What should i do? I suppose that any other developer who is not member of this repo will face the  same problems. (is this open-source?) 

About docker push of latest/current version: I have no privileges to push that images at all in openintegrationhub repo. So If i return any push, build will just fail.

About gcloud  installation and k8s. At the moment (in master branch) script just installs gcloud cli and does nothing. So what i've commented does not changes anything in build pipeline.  

So there is one more option: I revert back all changes in build pipeline, but we ignore broken build when merging PR (and repo settings should be adjusted for this)   

> We should go with the Community related to Kubernetes in our Github repo. So yaml is prefered Format here. Also a part of the DevOps Guideline which waiting in PR at the moment.

Ok. I understand this argument.

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js]: Maybe more meaningful error message? How about custom http errors?

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js]: More meaningful error message would be nice.

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js]: I propose to use `res.json` instead of `res.set` + `res.end`.

File: scripts/build.js
Comment (drobiazko) [reply scripts/build.js]: 👍 for the open source argument @anton-kotenko. ATM it looks like @falkvoigt is the only one enjoying the privilege to build the project successfully. This sucks big time and not how an open source project must work.

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js]: we need to show some message to a user

File: scripts/build.js
Comment (falkvoigt) [reply scripts/build.js]: We need to give permission for you or we make a condition related on Project name or id.

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js:81]: We need some helper in order to generate this kind of default message structure.

File: scripts/build.js
Comment (drobiazko) [reply scripts/build.js]: @falkvoigt this is an open source project. You would need to give the permissions to everybody on the globe.

Comment (falkvoigt) [reply scripts/build.js]: I can understand the fact with lag on permissions. But is it Open Source to disable Project related stuff locally and try to push it? And destroy the Pipeline which is only for OIH and not someone who wants to fork it?

it crashes on you side because you try to use Project internals. Normaly that is not the case when useing the OIH environment and not the forked one.

Comment (falkvoigt) [reply scripts/build.js]: not for every thing but nearly yes. I would restrict direct commiting to master. the other stuff i'm with you.

File: services/communication-router/src/HttpApi.js
Comment (pnedelko) [inline services/communication-router/src/HttpApi.js:101]: Is this property required?

Comment (pnedelko) [reply services/communication-router/src/HttpApi.js:101]: Is it used somewhere afterwards?

File: services/flows-operator/index.js
Comment (pnedelko) [inline services/flows-operator/index.js:8]: why `console.error` instead of `logger`?

File: scripts/build.js
Comment (falkvoigt) [reply scripts/build.js]: @drobiazko i vote for open branch for every one restrict access for master. I'm not the github admin to do so. I just comment that the PR will brake pipeline on Project side.

Comment (anton-kotenko) [reply scripts/build.js]: About open-source: At  this moment external developer is not able to create PR and successfully pass tests.  So it's a problem of test+build pipeline or general work process organization. 
I don't think that my solution of the problem is the best, but problem exists. And we should do something with this. 
So we have a question: how to avoid all the internals (build, docker push, gcloud) for external users.

Comment (falkvoigt) [reply scripts/build.js]: i already answered that.
2 different approches:
1. git branch create permission to every one and don't try to PR from a fork
2. make a condition based on Projectname/id to the push and google stuff

File: services/flows-operator/src/FlowOperator.js
Comment (pnedelko) [inline services/flows-operator/src/FlowOperator.js:156]: Typo: Inedx => Index

**Comment by falkvoigt** [comment]
hi i just had a call with @drobiazko  related to that.
we just merge the PR and i will care on the Contitions related to the Pipeline

File: services/flows-operator/src/FlowOperator.js
Comment (pnedelko) [inline services/flows-operator/src/FlowOperator.js]: This like and the line above is a C&P

File: services/flows-operator/src/HttpApi.js
Comment (pnedelko) [inline services/flows-operator/src/HttpApi.js]: `res.json` is better

File: services/flows-operator/src/HttpApi.js
Comment (pnedelko) [inline services/flows-operator/src/HttpApi.js]: Would be nice to respond with some message, as well as with different http codes.

File: services/scheduler/src/Scheduler.js
Comment (pnedelko) [inline services/scheduler/src/Scheduler.js]: I'm not aware of the style guides, but seems like `'` should be here instead of `"`.

**Comment by anton-kotenko** [comment]
@falkvoigt  I've fixed .travis.yaml to run tests but skip builds for forked repos, and revert back  scripts/*
Is this ok?

File: services/scheduler/src/Scheduler.js
Comment (pnedelko) [inline services/scheduler/src/Scheduler.js]: Naming convention: would be nice to stick with one of the following: `node` or `step`. I would prefer `node`.

**Comment by falkvoigt** [comment]
Yes this is a great solution to the Problem.
Even better than my condition in the script for sure.

File: services/scheduler/src/Scheduler.js
Comment (pnedelko) [inline services/scheduler/src/Scheduler.js]: again codestyle: is `"` ok here?

File: services/scheduler/src/Scheduler.js
Comment (pnedelko) [inline services/scheduler/src/Scheduler.js]: Would be nice to move magic number to some constant.

Comment (pnedelko) [reply services/scheduler/src/Scheduler.js]: And how about `cron` flows?

File: src/lib/src/K8sService.js
Comment (pnedelko) [inline src/lib/src/K8sService.js]: Looks to me more like `init`, cause it doesn't really establish any connection or something like that.

**Review by pnedelko — CHANGES_REQUESTED** [review]

File: services/scheduler/src/Scheduler.js
Comment (anton-kotenko) [reply services/scheduler/src/Scheduler.js]: cron flows are not implemented atm

**Review by pnedelko — APPROVED** [review]

**Pull request merged**
